packages/cardano-services/src/Program/services/ogmios.ts

```diff
@@ -0,0 +1,151 @@
+                );
+              ogmiosProvider = ogmiosTxSubmitProvider({ host: record.name, port: record.port });
+              return await ogmiosProvider.submitTx(args);
+            }
+            throw error;
+          });
+      }
+      // Bind if it is a function, no intercept operations
+      if (typeof ogmiosProvider[prop as keyof TxSubmitProvider] === 'function') {
+        const method = ogmiosProvider[prop as keyof TxSubmitProvider] as any;
+        return method.bind(ogmiosProvider);
+      }
+    }
+  });
+};
+
+export const getOgmiosTxSubmitProvider = async (
+  dnsResolver: DnsResolver,
+  logger: Logger,
+  options?: CommonProgramOptions
+): Promise<TxSubmitProvider> => {
+  if (options?.ogmiosSrvServiceName)
+    return ogmiosTxSubmitProviderWithDiscovery(dnsResolver, logger, options.ogmiosSrvServiceName);
+  if (options?.ogmiosUrl) return ogmiosTxSubmitProvider(urlToConnectionConfig(options?.ogmiosUrl));
+  throw new MissingProgramOption(ServiceNames.TxSubmit, [
+    CommonOptionDescriptions.OgmiosUrl,
+    CommonOptionDescriptions.OgmiosSrvServiceName
+  ]);
+};
+
+/**
+ * Creates a extended OgmiosCardanoNode instance :
+ * - use passed srv service name in order to resolve the port
+ * - make dealing with failovers (re-resolving the port) opaque
+ * - use exponential backoff retry internally with default timeout and factor
+ * - intercept 'initialize' operation and handle connection errors on initialization
+ * - intercept 'eraSummaries' operation and handle connection errors runtime
+ * - all other operations are bind to pool object withoud modifications
+ *
+ * @returns OgmiosCardanoNode instance
+ */
+export const ogmiosCardanoNodeWithDiscovery = async (
+  dnsResolver: DnsResolver,
+  logger: Logger,
+  serviceName: string
+): Promise<OgmiosCardanoNode> => {
+  const { name, port } = await dnsResolver(serviceName!);
+  let ogmiosCardanoNode = new OgmiosCardanoNode({ host: name, port }, logger);
+
+  return new Proxy<OgmiosCardanoNode>({} as OgmiosCardanoNode, {
+    get(_, prop) {
+      if (prop === 'then') return;
+      if (prop === 'initialize') {
+        return () =>
+          ogmiosCardanoNode.initialize().catch(async (error) => {
+            if (isConnectionError(error)) {
+              const record = await dnsResolver(serviceName!);
+              logger.info(`DNS resolution for Ogmios service, resolved with record: ${JSON.stringify(record)}`);
+              await ogmiosCardanoNode
+                .shutdown?.()
+                .catch((error_) =>
+                  logger.warn(`Ogmios cardano node failed to shutdown after DNS resolution: ${error_}`)
+                );
+              ogmiosCardanoNode = new OgmiosCardanoNode({ host: record.name, port: record.port }, logger);
+              return await ogmiosCardanoNode.initialize();
+            }
+            throw error;
+          });
+      }
+      if (prop === 'eraSummaries') {
```

Comment:
I think we also need to override `systemStart`. Can probably add `'initialize'` as well to dedupe some code.

```suggestion
      const isCardanoNodeOperation = (prop: unknown): prop is 'eraSummaries' | 'systemStart' => ['eraSummaries', 'systemStart'].includes(prop);
      if (isCardanoNodeOperation(prop)) {
```

Reply:
I initially checked that [systemStart](https://github.com/input-output-hk/cardano-js-sdk/blob/08d9ed2b6aa20cf4df2a063f046f4e5ca28c6bd5/packages/ogmios/src/CardanoNode/OgmiosCardanoNode.ts#L65) has no real usage and refs except in the tests. The `stateQueryClient.systemStart()` is directly used within [eraSummaries](https://github.com/input-output-hk/cardano-js-sdk/blob/08d9ed2b6aa20cf4df2a063f046f4e5ca28c6bd5/packages/ogmios/src/CardanoNode/OgmiosCardanoNode.ts#L51-L63) method, that's why didn't override Cardano Node one. Shall we override it then just in case? @mkazlauskas

Reply:
Yes please, if it gets used then it'll be a bug. Or make it private.

Reply:
_> Can probably add 'initialize' as well_
It differs from the other operations, as there it needs to invoke first (`initialize`) and then the given operation (`systemStart` or `eraSummaries`), where if `initialize` fails we return it after DNS resolution

Reply:
I see, maybe you could hoist [these](https://github.com/input-output-hk/cardano-js-sdk/blob/e90f99d6ad7ac47669c50b388b2a1b137d5de3aa/packages/cardano-services/src/Program/services/ogmios.ts#L99-L106) in a function like `recreateCardanoNode` then

Reply:
The changes are applied under: 629e967ee41670107a9f2fb26e0328366fadfa51